admin/jobs/river/river.go

```diff
@@ -0,0 +1,145 @@
+package river
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/rilldata/rill/admin"
+	"github.com/rilldata/rill/admin/jobs"
+	"github.com/rilldata/rill/runtime/pkg/observability"
+	"github.com/riverqueue/river"
+	"github.com/riverqueue/river/riverdriver/riverpgxv5"
+	"github.com/riverqueue/river/rivermigrate"
+	"github.com/robfig/cron/v3"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+)
+
+var (
+	tracer              = otel.Tracer("github.com/rilldata/rill/admin/jobs/river")
+	meter               = otel.Meter("github.com/rilldata/rill/admin/jobs/river")
+	jobLatencyHistogram = observability.Must(meter.Int64Histogram("job_latency", metric.WithUnit("ms")))
+)
+
+type RiverClient struct {
+	dbPool      *pgxpool.Pool
+	riverClient *river.Client[pgx.Tx]
+}
+
+func New(ctx context.Context, dsn string, adm *admin.Service) (jobs.Client, error) {
+	dbPool, err := pgxpool.New(ctx, dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	tx, err := dbPool.Begin(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = tx.Rollback(ctx) }()
+
+	migrator := rivermigrate.New(riverpgxv5.New(dbPool), nil)
+
+	res, err := migrator.MigrateTx(ctx, tx, rivermigrate.DirectionUp, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	err = tx.Commit(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, version := range res.Versions {
+		adm.Logger.Info("river database migrated", zap.String("direction", string(res.Direction)), zap.Int("version", version.Version))
+	}
+
+	workers := river.NewWorkers()
+	// NOTE: Register new job workers here
+	river.AddWorker(workers, &ValidateDeploymentsWorker{admin: adm})
+	river.AddWorker(workers, &ResetAllDeploymentsWorker{admin: adm})
+
+	periodicJobs := []*river.PeriodicJob{
+		// NOTE: Add new periodic jobs here
+		newPeriodicJob(&ValidateDeploymentsArgs{}, "* */6 * * *", true),
+	}
+
+	riverClient, err := river.NewClient(riverpgxv5.New(dbPool), &river.Config{
+		Queues: map[string]river.QueueConfig{
+			river.QueueDefault: {MaxWorkers: 100},
+		},
+		Workers:      workers,
+		PeriodicJobs: periodicJobs,
+	})
```

Comment:
Should we also set:
- `JobTimeout` – we have some batch jobs that likely will run for more than a minute (the default). Should we maybe disable the timeout entirely?
- `Logger` – looks like it will default to STDOUT at above warn level. Might be nice to wire this up to our Zap logger so we get structured logs in the same format as our other logs.

Reply:
might be good to add an error handler as well like [here](https://github.com/rilldata/rill/pull/5440/files#diff-aacc074d2e424275f2c45d80c99ece23ee71b9c6b456f07d1e47b0a41709e057R702) to properly log job failures. I think we should add a timeout of 10 minute. For refs I used these configs - https://github.com/rilldata/rill/pull/5440/files#diff-1ff6c7660e144613f9efda83929a9eefca974d681429e2a8dfc46fb11aa7fdfeR43

Reply:
- I've added a global `JobTimeout` of 1 hour, the timeout can also be overriden on the worker level for very long running jobs
- @pjain1 good idea with the error handler, I've copied over the `ErrorHandler` from your branch
- The logger is a bit tricky since it expect a `slog` logger and it seems like we would need something like https://github.com/samber/slog-zap to get it to work with zap. Should we give that library a try?

Reply:
@kaspersjo How about just using this adapter from the core Zap repo: https://github.com/uber-go/zap/blob/master/exp/zapslog/example_test.go?

Reply:
Added a logger using the handler from Zap core